pkg/edgeview/src/network.go

```diff
@@ -0,0 +1,1052 @@
+						isApp = true
+						appStr = " (appNum " + strconv.Itoa(appN) + ")"
+					}
+				}
+			}
+			if !isApp && querytype == "app" {
+				continue
+			}
+		}
+		olines = l + appStr
+		if subStr != "" {
+			olines = strings.ReplaceAll(olines, subStr, "\033[0;93m"+subStr+"\033[0m")
+		}
+		fmt.Printf("%s", olines)
+		if i%20 == 0 {
+			closePipe(true)
+		} else {
+			fmt.Println()
+		}
+	}
+}
+
+func runmDNS(subStr string) {
+	if directQuery {
+		fmt.Printf("mdns is not supported in edge-view ssh mode\n")
+		return
+	}
+
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		fmt.Printf("get interface error: %v\n", err)
+		return
+	}
+
+	var intfname, serv, serviceStr string
+	if strings.Contains(subStr, "/") {
+		substrings := strings.Split(subStr, "/")
+		if len(substrings) != 2 {
+			fmt.Printf("mdns parameter is in the form interface/service\n")
+			return
+		}
+		intfname = substrings[0]
+		serv = substrings[1]
+	} else {
+		intfname = subStr
+	}
+	var ifs []net.Interface
+	for _, intf := range ifaces {
+		//fmt.Printf("interface %+v\n", intf)
+		if intfname == "" {
+			if strings.HasPrefix(intf.Flags.String(), "up|") {
+				ifs = append(ifs, intf)
+			}
+		} else if intf.Name == intfname {
+			ifs = append(ifs, intf)
+			break
+		}
+	}
+
+	if serv == "" {
+		serviceStr = "_workstation._tcp"
+	} else {
+		serviceStr = "_" + serv + "._tcp"
+	}
+
+	var port []string
+	for _, p := range ifs {
+		port = append(port, p.Name)
+	}
+	printTitle(fmt.Sprintf("query mDNS service %s, on intfs %v\n", serviceStr, port), colorCYAN, true)
+
+	ifOption := zeroconf.SelectIfaces(ifs)
+	ipOption := zeroconf.SelectIPTraffic(zeroconf.IPv4)
+	resolver, err := zeroconf.NewResolver(ipOption, ifOption)
+	if err != nil {
+		fmt.Printf("queryService: Failed to initialize resolver: %v", err)
+		return
+	}
+
+	mctx, cancel := context.WithTimeout(context.Background(), time.Second*time.Duration(10))
+	defer cancel()
+
+	entries := make(chan *zeroconf.ServiceEntry)
+	go func(results <-chan *zeroconf.ServiceEntry) {
+		for entry := range results {
+			fmt.Printf("  - %+v\n", entry)
+		}
+	}(entries)
+
+
+	err = resolver.Browse(mctx, serviceStr, "local", entries)
+	if err != nil {
+		fmt.Printf("mdns resolver error %v", err)
+		return
+	}
+	<-mctx.Done()
+}
+
+func runTCPDump(subStr string) {
+	if !strings.Contains(subStr, "/") {
+		fmt.Printf("need to have intf name separated by slash\n")
+		return
+	}
+	err := addPackage("tcpdump", "tcpdump")
+	if err != nil {
+		return
+	}
+
+	subs := strings.SplitN(subStr, "/", 2)
+	intf := subs[0]
+	var timeValue string
+	if timeout != "" {
+		timeSec, err := strconv.Atoi(timeout)
+		if err != nil {
+			fmt.Printf("time option has to be seconds: %v\n", err)
+			return
+		}
+		if timeSec > 120 {
+			timeValue = "120"
+			fmt.Printf("time value for tcpdump maximum is 120 seconds\n\n")
+		} else {
+			timeValue = timeout
+		}
+	} else {
+		timeValue = "60"
+	}
+	cmd := "timeout " + timeValue + " tcpdump -i " + intf + " " + subs[1] + " -c 100"
```

Comment:
Note that this arbitrary subs[1] argument allows the user to pass in e.g., "; reboot". The same is actually true for the intf argument - "tcpdump/eth1;reboot" will result in a reboot.

If we really want to restrict what the user can do we need a different approach with only specific arguments and those are checked to not include characters which are special to the shell, OR we avoid using the shell and instead parse the arguments and construct the argv array and pass that to execv(). Latter might be a reasonable compromise to reduce the amount of mischief which can be done using the shell.

Reply:
@naiming-zededa can you check that intf and subs[1] does not contain anything but letters and digits? Othwewise I can use tcpdump to do any shell command.

Reply:
added the check for against all the up and valid interface name, and check for ';' string in the sub[1] in new commit

Reply:
@eriknordmark  committed the code to check 'intf' against all the 'UP' interfaces name, and check for ';' string in subs[1]

Reply:
But ';' is only one of lots of characters which can be used to fool the shell. '|' is another one. Hence the request to only allow alphanumeric characters.

Basically all calls to runCmd need to have anything which can be possible controllerd by the user sanitized.
And to make this easier to review it would be a good idea to replace all of the unnecessary runcmd which invoke ls, cat, ip use the golang functions (ReadFile, Readdir, netlink) so that one can look more carefully at the ones which invoke tcpdump, ping, nslookup, grep, etc.

Reply:
@eriknordmark ok. but since this 'edge-view' also have the mode for 'ssh' to gather the data from device, that part can not be in golang functions. let me go through the list, and remove certain options (e.g. remove the above sub[1] completely) to have commands with less flexibility and more security way.

Reply:
But as I asked in the other comment from today, isn't this ssh to handle old devices where EVE-OS has no edge-view support? 
While that might be useful to have in our back pocket for the cases where we have ssh access, it does not have to be part of the Edge-view which we put in EVE-OS; there we will run the commands on the device itself and will access the files and directories etc.

Reply:
I can remove that, then the ssh-mode can not be supported in the future unless we keep different repos, currently e.g. Ruslan's test job for traffic is using the 'edge-view' ssh-mode.